drivers/spi/spi_shell.c

```diff
@@ -4,54 +4,275 @@
+	(COND_CODE_1(DT_NODE_HAS_STATUS_OKAY(DT_BUS(node_id)), \
+	(fn), ())), ())
+
+/* Create specified number of empty structs, separated by ',' */
+#define _EMPTY_STRUCT_INST(idx, list) {0}
+#define CREATE_NUM_EMPTY_STRUCTS(num) LISTIFY(num, _EMPTY_STRUCT_INST, (,))
+
+/* Struct representing either a spi bus or a spi device. */
+struct spi_thing {
+	/* Device name. Either a spi bus or a spi device. */
+	const char *name;
+	struct spi_dt_spec spec;
+};
+
+/* Struct used to map a label to a name of the associated spi_thing. */
+struct map {
+	/* Either nodelabel or device name of either spi bus or spi device. */
+	const char *label;
+	/* Device name. This is the same as the name of the associated spi_thing struct. */
+	const char *name;
+};
+
+#define INST_SPI_THING_AS_SPI_DEV(node_id)                                                         \
+	{                                                                                          \
+		.name = DEVICE_DT_NAME(node_id),                                                   \
+		.spec = SPI_DT_SPEC_GET(node_id, 0, 0),                                            \
+	},
+
+#define INST_SPI_THING_AS_SPI_DEV_AS_SPI_BUS(dev)                                                  \
+	(struct spi_thing)                                                                         \
+	{                                                                                          \
+		.name = dev->name,                                                                 \
+		.spec = {                                                                          \
+			.bus = dev,                                                                \
+			.config =                                                                  \
+				{                                                                  \
+					.frequency = 1000000,                                      \
+					.operation = SPI_OP_MODE_MASTER | SPI_WORD_SET(8),         \
+				},                                                                 \
+		},                                                                                 \
+	}
+
+#define INST_MAP(nodelabel, node_id)                                                               \
+	{.label = STRINGIFY(nodelabel), .name = DEVICE_DT_NAME(node_id)},
+
+#define INST_MAP_FROM_NODE_ID(node_id)                                                             \
+	{.label = DEVICE_DT_NAME(node_id), .name = DEVICE_DT_NAME(node_id)},
+
+/* Instantiate spi thing struct from node_id, if that is a spi device.  */
+#define INST_ALL_SPI_DEVICES_AS_SPI_THINGS(node_id)                                                \
+	RUN_FN_ON_SPI_DEVICE(node_id, INST_SPI_THING_AS_SPI_DEV(node_id))
+
+/* List of spi things. At compile time we instantiate structs for all spi devices.
+ * Additional empty space is reserved for the spi buses, which are added in spi_buses_init at
+ * runtime.
+ */
+static struct spi_thing spi_things[] = {
```

Comment:
`spi_device_registry` or `spi_shell_devices`?

Reply:
`spi_shell_devices`, since it's type is `spi_shell_device`.